designs/karpenter-dra-kwok-driver.md

```diff
@@ -0,0 +1,135 @@
+      # ... (6 more devices for total of 8)
+```
+
+**How it works**:
+1. **Test author defines NodeOverlay configuration**: "g5.48xlarge KWOK nodes should have 8x fake H100 GPUs" via ResourceSlices
+2. **Driver watches for KWOK nodes**: When Karpenter creates a KWOK node with `instance-type: g5.48xlarge`
+3. **NodeOverlay match found**: Driver checks for NodeOverlay with embedded ResourceSlice objects, finds matching configuration
+4. **Driver creates ResourceSlice**: Acts as fake DRA driver using embedded ResourceSlice objects from NodeOverlay
+5. **Scheduler sees configured devices**: ResourceSlices with fake devices become available for DRA pod scheduling
+6. **Test validation**: Validates that the driver correctly provides DRA resources and enables successful pod scheduling
+
+### Case 2: ConfigMap Fallback Configuration
+Tests **DRA resource provisioning when no NodeOverlay configuration is found**. When the driver cannot find matching NodeOverlay resources with DRA annotations for a KWOK node, it falls back to ConfigMap-based device configuration. This validates that the driver can handle nodes that weren't pre-configured with DRA resources in NodeOverlay and ensures consistent DRA device availability across different provisioning scenarios.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: kwok-dra-config
+data:
+  config.yaml: |
+    driver: "fake-gpu.kwok.x-k8s.io"
+    mappings:
+    - name: "h100-nodes"
+      nodeSelector:
+        matchLabels:
+          node.kubernetes.io/instance-type: "g5.48xlarge"
+          kwok.x-k8s.io/node: "fake"
+      resourceSlice:
+        devices:
+        - name: "nvidia-h100"
+          count: 8
+          attributes:
+            memory: "80Gi"
+            compute-capability: "9.0" 
+            device-type: "gpu"
+            vendor: "nvidia"
+    - name: "fpga-nodes"
+      nodeSelector:
+        matchLabels:
+          node.kubernetes.io/instance-type: "f1.2xlarge"
+          kwok.x-k8s.io/node: "fake"
+      resourceSlice:
+        devices:
+        - name: "xilinx-u250"
+          count: 1
+          attributes:
+            memory: "16Gi"
+            device-type: "fpga"
+            vendor: "xilinx"
+```
+
+**How it works**:
+1. **Test author defines ConfigMap configuration**: "g5.48xlarge KWOK nodes should have 8x fake H100 GPUs when no NodeOverlay is found"
+2. **Driver watches for KWOK nodes**: When Karpenter creates a KWOK node with `instance-type: g5.48xlarge`
+3. **No NodeOverlay match found**: Driver checks for NodeOverlay with embedded ResourceSlice objects, finds none, falls back to ConfigMap
+4. **Driver creates ResourceSlice**: Acts as fake DRA driver using ConfigMap configuration
+5. **Scheduler sees configured devices**: ResourceSlices with fake devices become available for DRA pod scheduling
+6. **Test validation**: Validates that the driver correctly provides DRA resources and enables successful pod scheduling
+
+## Possible Directory Structure
```

Comment:
I think we'll likely want to have a separate "project directory", similar to the existing kwok directory, to host this controller. The project in that directory can still use upstream Karpenter as a library if it makes sense to have some shared functionality, but I don't think it makes sense to include the business logic in the main Karpenter codebase. Let me know what you think.